<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <link href='https://fonts.googleapis.com/css?family=Chivo:900' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-dark.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <title>Analysis of epigenetic signals captured by fragmentation patterns of cell-free DNA by shendurelab</title>
  </head>

  <body>
    <div id="container">
      <div class="inner">

        <header>
          <h1>Analysis of epigenetic signals captured by fragmentation patterns of cell-free DNA</h1>
          <h2>Scripts for the study &quot;Cell-free DNA Comprises an In Vivo Nucleosome Footprint that Informs Its Tissues-Of-Origin&quot;</h2>
        </header>

        <section id="downloads" class="clearfix">
          <a href="https://github.com/shendurelab/cfDNA/zipball/master" id="download-zip" class="button"><span>Download .zip</span></a>
          <a href="https://github.com/shendurelab/cfDNA/tarball/master" id="download-tar-gz" class="button"><span>Download .tar.gz</span></a>
          <a href="https://github.com/shendurelab/cfDNA" id="view-on-github" class="button"><span>View on GitHub</span></a>
        </section>

        <hr>

        <section id="main_content">
          <h1>
<a id="analysis-of-epigenetic-signals-captured-by-fragmentation-patterns-of-cell-free-dna" class="anchor" href="#analysis-of-epigenetic-signals-captured-by-fragmentation-patterns-of-cell-free-dna" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Analysis of epigenetic signals captured by fragmentation patterns of cell-free DNA</h1>
<p>The following sections provide details for the analyses performed in:</p>
<p>Snyder MW, Kircher M, Hill AJ, Daza RM, Shendure J. Cell-free DNA Comprises
an In Vivo Nucleosome Footprint that Informs Its Tissues-Of-Origin. Cell. 2016 Jan
14;164(1-2):57-68. doi: 10.1016/j.cell.2015.11.050. PubMed PMID: <a href="http://www.ncbi.nlm.nih.gov/pubmed/26771485">26771485</a></p>
<p><em>All scripts and binaries are provided as is, without any warrenty and for use at your own risk. This is not the release of a software package. We are only providing this information and code in addition to a description of methods for making it easier to reproduce our analyses. We are <strong>not</strong> providing any support for these scripts.</em></p>
<h2>
<a id="versions-and-dependencies" class="anchor" href="#versions-and-dependencies" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Versions and dependencies</h2>
<p>Our scripts largely depend on Python 2, the <a href="https://pysam.readthedocs.io/en/latest/api.html">pysam</a>, <a href="https://pypi.org/project/bx-python/">bx</a> and <a href="https://numpy.org/">numpy</a> libraries. Here a list of versions that we used:</p>
<pre><code>python/2.7.3
numpy/1.7.0
pysam/0.7.5
bx-python/0.6.0
</code></pre>
<p>We were also using <a href="https://cran.r-project.org/">R 3.1.0</a>, <a href="http://hgdownload.cse.ucsc.edu/admin/exe/">UCSC binaries</a> for working with bigWig and bigBed files and tools like tabix and samtools from <a href="http://www.htslib.org/">HTSlib</a>.</p>
<p>Please consider using a software management tool like conda to install the respective packages. You can try more recent versions of the packages, but please keep in mind that Python 3 cannot be used to interpret Python 2.7 code.</p>
<h3>
<a id="samtools-version-used" class="anchor" href="#samtools-version-used" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Samtools version used</h3>
<p>Please note that a samtools binary is included with these scripts. Among other things, this samtools binary allows filtering reads based on insert size/read length. This is an early version of the samtools branch released on <a href="https://github.com/mpieva/samtools-patched">https://github.com/mpieva/samtools-patched</a>. For samtool calls that are not filtering for read length/insert size, other (more recent) versions of samtools might be used. Please note though that parameters might be named differently and that the ASCII encoding of read filters is also special to the samtools version that we used.</p>
<h2>
<a id="primary-data-processing-of-sequencing-data" class="anchor" href="#primary-data-processing-of-sequencing-data" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Primary data processing of sequencing data</h2>
<p>Barcoded paired-end (PE) sequencing data was split allowing up to one substitution in the barcode sequence. Fragments shorter than or equal to the read length were consensus-called and adapter-trimmed. Remaining consensus single-end reads (SR) and the individual PE reads were aligned to the human reference genome (GRCh37, 1000 Genomes phase 2 technical reference, ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/phase2_reference_assembly_sequence/) using the ALN algorithm in BWA v0.7.10 (<a href="http://www.ncbi.nlm.nih.gov/pubmed/20080505">Li and Durbin, 2010</a>). PE reads were further processed with BWA SAMPE to resolve ambiguous placement of read pairs or to rescue missing alignments by a more sensitive alignment step around the location of one placed read end. Aligned SR and PE data were stored in BAM format using the samtools API (<a href="http://www.ncbi.nlm.nih.gov/pubmed/19505943">Li et al., 2009</a>). BAM files for each sample were merged across lanes and sequencing runs. BAM files for each sample were deposited in the NCBI Gene Expression Omnibus (GEO) with accession GSE71378 (<a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE71378">http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE71378</a>). Note that GEO is distributing BAM files through SRA (<a href="http://www.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?study=SRP061633">http://www.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?study=SRP061633</a> and <a href="http://www.ncbi.nlm.nih.gov/Traces/study/?acc=SRP061633">http://www.ncbi.nlm.nih.gov/Traces/study/?acc=SRP061633</a>). <em>Update:</em> SRA has recently converted the provided BAM files to SRA format. You can use the SRA tools to convert back. We are providing a copy of the original BAM files <a href="https://kircherlab.bihealth.org/download/cfDNA/">here</a>.</p>
<h2>
<a id="simulated-reads-and-nucleotide-frequencies" class="anchor" href="#simulated-reads-and-nucleotide-frequencies" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Simulated reads and nucleotide frequencies</h2>
<p>Aligned sequencing data was simulated (SR if shorter than 45 bp, PE 45 bp otherwise) for all major chromosomes of the human reference (GRC37h). Kmer nucleotide frequencies (k=2 is presented in the manuscript) were determined from real data on both fragment ends and both strand orientations (<code>referenceKMers.py</code>), and for the reference genome on both strands (<code>BAM2FragKMers.py</code>). The insert size distribution of the real data was extracted for the 1-500 bp range (<code>BAM_RG_Length.py --noRG</code>). Reads were simulated procedurally (implemented in <code>simulate_reads.py</code>): at each step (i.e., at least once at each genomic coordinate, depending on desired coverage), (1) the strand is randomly chosen, (2) the ratio of the kmer frequency in the real data to that in the reference sequence is used to randomly decide whether the initiating kmer is considered, (3) an length is sampled from the insert size distribution, and (4) the frequency ratio of the terminal kmer is used to randomly decide whether the generated alignment is reported. The simulated coverage was matched to that of the original data after PCR duplicate removal (by adjusting how often sequences are sampled at each position).</p>
<div class="highlight highlight-source-shell"><pre><span class="pl-c"><span class="pl-c">#</span> Extract kmers (here 2mers) from reference genome </span>
./referenceKMers.py -k 2 -r <span class="pl-s"><span class="pl-pds">'</span><span class="pl-pds">'</span></span> -o grch37_regChroms_2mers.tsv

<span class="pl-c"><span class="pl-c">#</span> Extact the length distribution from the sample BAM, keep only the 1-500bp range </span>
./BAM_RG_Length.py --noRG -p tmp_outfile BAMFILE.bam
head -n 501 tmp_outfile_ <span class="pl-k">&gt;</span> <span class="pl-smi">${SAMPLE}</span>_lenDist.tsv

<span class="pl-c"><span class="pl-c">#</span> Extract kmers (here 2mers) of left and right read ends from the SAMPLE file</span>
./BAM2FragKMers.py --kmerLE=2 --kmerRE=2 -v -r <span class="pl-s"><span class="pl-pds">'</span>ALL<span class="pl-pds">'</span></span> BAMFILE.bam --outfileLE=<span class="pl-smi">${SAMPLE}</span>_left_2mer --outfileRE=<span class="pl-smi">${SAMPLE}</span>_right_2mer

<span class="pl-c"><span class="pl-c">#</span> Run simulation for each "regular" chromosome (1..22, X, Y) -- should be run in parallel</span>
<span class="pl-k">for</span> <span class="pl-smi">i</span> <span class="pl-k">in</span> <span class="pl-s"><span class="pl-pds">$(</span>head -n 24 grch37_1000g_phase2/whole_genome.fa.fai <span class="pl-k">|</span> awk <span class="pl-s"><span class="pl-pds">'</span>{ print $1":1-"$2 }<span class="pl-pds">'</span></span><span class="pl-pds">)</span></span><span class="pl-k">;</span> <span class="pl-k">do</span> \
  ./simulate_reads.py -v -d <span class="pl-smi">${SAMPLE}</span>_lenDist.tsv --fwdKMerGenome=grch37_regChroms_2mers.tsv --revKMerGenome=grch37_regChroms_2mers.tsv --fwdPKMers=<span class="pl-smi">${SAMPLE}</span>_left_2mer_f.tsv --fwdMKMers=<span class="pl-smi">${SAMPLE}</span>_left_2mer_r.tsv --revPKMers=<span class="pl-smi">${SAMPLE}</span>_right_2mer_f.tsv --revMKMers=<span class="pl-smi">${SAMPLE}</span>_right_2mer_r.tsv -s 20  -r <span class="pl-cce">\'</span><span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$i</span><span class="pl-pds">"</span></span><span class="pl-cce">\'</span> -o <span class="pl-smi">${SAMPLE}</span>_2mer_sim_chr<span class="pl-s"><span class="pl-pds">$(</span> <span class="pl-c1">echo</span> <span class="pl-smi">$i</span> <span class="pl-k">|</span> cut -f 1 -d<span class="pl-s"><span class="pl-pds">'</span>:<span class="pl-pds">'</span></span><span class="pl-pds">)</span></span>.bam<span class="pl-k">;</span> \
<span class="pl-k">done</span>

<span class="pl-c"><span class="pl-c">#</span> Combine chromosome files, index and regenerate some stats</span>
./samtools merge -u <span class="pl-smi">${SAMPLE}</span>_chr<span class="pl-k">*</span>.bam <span class="pl-k">|</span> ./samtools sort - <span class="pl-smi">${SAMPLE}</span>_allChrom
./samtools index <span class="pl-smi">${SAMPLE}</span>_allChrom.bam
./samtools flagstat <span class="pl-smi">${SAMPLE}</span>_allChrom.bam <span class="pl-k">&gt;</span> <span class="pl-smi">${SAMPLE}</span>_allChrom.bam_stats
./samtools view -F _2 <span class="pl-smi">${SAMPLE}</span>_allChrom.bam <span class="pl-k">|</span> cut -f 3 <span class="pl-k">|</span> uniq -c <span class="pl-k">&gt;</span> <span class="pl-smi">${SAMPLE}</span>_allChrom.byChromCounts.txt
./BAM2FragmentationPatterns.py -r ALL -o <span class="pl-smi">${SAMPLE}</span>_allChrom.fragpatterns.txt <span class="pl-smi">${SAMPLE}</span>_allChrom.bam
./BAM_RG_Length.py --noRG -p <span class="pl-smi">${SAMPLE}</span>_allChrom.length <span class="pl-smi">${SAMPLE}</span>_allChrom.bam</pre></div>
<h2>
<a id="analysis-of-nucleotide-composition-of-167-bp-fragments" class="anchor" href="#analysis-of-nucleotide-composition-of-167-bp-fragments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Analysis of nucleotide composition of 167 bp fragments</h2>
<p>Fragments with inferred lengths of exactly 167 bp were filtered within samples to remove duplicates. Dinucleotide frequencies were calculated in a strand-aware manner, using a sliding 2 bp window and reference alleles at each position, beginning 50 bp upstream of one fragment endpoint and ending 50 bp downstream of the other endpoint. Observed dinucleotide frequencies at each position were compared to expected dinucleotide frequencies determined from a set of simulated reads reflecting the same cleavage biases calculated in a library-specific manner.</p>
<p>See the folder <code>nucleotide_composition</code> for more details.</p>
<h2>
<a id="coverage-fragment-endpoints-and-windowed-protection-scores" class="anchor" href="#coverage-fragment-endpoints-and-windowed-protection-scores" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Coverage, fragment endpoints, and windowed protection scores</h2>
<p>Fragment endpoint coordinates were extracted from BAM files with the SAMtools API. Both outer alignment coordinates of PE data were extracted for properly paired reads. Both end coordinates of SR alignments were extracted when PE data was collapsed to SR data by adapter trimming. A fragment’s coverage is defined as all positions between the two (inferred) fragment ends, inclusive of endpoints. We define the Windowed Protection Score (WPS) of a window of size k as the number of molecules spanning the window minus those with an endpoint within the window. We assign the determined WPS to the center of the window. For 35-80 bp fragments (short fraction, S-WPS), k=16; for 120-180 bp fragments (long fraction, L-WPS), k=120. We store coverage, read starts and WPS in gzip-compressed WIG files. We used wigToBigWig from the UCSC tools (<a href="http://hgdownload.cse.ucsc.edu/admin/exe/">http://hgdownload.cse.ucsc.edu/admin/exe/</a>) for converting these to BigWig (bw) for visualization in genome viewers.</p>
<p>Running the extraction of coverage, read starts and WPS for the long fraction:</p>
<div class="highlight highlight-source-shell"><pre>./samtools view -u -m 120 -M 180 BAMFILE.bam <span class="pl-smi">$EXTREGION</span> <span class="pl-k">|</span> ./FilterUniqueBAM.py -p <span class="pl-k">|</span> ./extractReadStartsFromBAM2Wig.py -p -r <span class="pl-smi">$REGION</span> -w 120 -c COVERAGE.wig.gz -n WPS.wig.gz -s STARTS.wig.gz</pre></div>
<p>Running the extraction of coverage, read starts and WPS for the short fraction:</p>
<div class="highlight highlight-source-shell"><pre>./samtools view -u -m 35 -M 80 BAMFILE.bam <span class="pl-smi">$EXTREGION</span> <span class="pl-k">|</span> ./FilterUniqueBAM.py -p <span class="pl-k">|</span> ./extractReadStartsFromBAM2Wig.py -p -r <span class="pl-smi">$REGION</span> -w 16 -c COVERAGE.wig.gz -n WPS.wig.gz -s STARTS.wig.gz</pre></div>
<p>Please note the variables <code>EXTREGION</code> and <code>REGION</code> above. When piping reads from a region, we might miss the forward read of a read pair and the provided scripts usually only extract information from the first read of a pair. Thus, <code>EXTREGION</code> should include additional bases around the region (i.e. 200bp or another value that guarantees that all read insert sizes are included; here 180bp and 80bp would be sufficient; e.g. if REGION is 1:1000-2000, EXTREGION should be 1:800-2200).</p>
<h2>
<a id="nucleosome-peak-calling" class="anchor" href="#nucleosome-peak-calling" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Nucleosome peak calling</h2>
<p>For nuclesome peak calling, the L-WPS is locally adjusted to a running median of zero in 1 kb windows and smoothed using a Savitzky-Golay filter (<a href="http://pubs.acs.org/doi/abs/10.1021/ac60214a047">Savitzky and Golay, 1964</a>) (window size 21, 2nd order polynomial). The L-WPS track is then segmented into above-zero regions (allowing up to 5 consecutive positions below zero). If the resulting region is 50-150 bp, we identify the median L-WPS value of that region and search for the maximum-sum contiguous window above the median. We report the start, end and center coordinates of this window as the “peak,” or local maximum of nucleosome protection. All calculations involving distances between peaks are based on these center coordinates. A score for each peak is determined as the distance between maximum value in the window and the average of the two adjacent L-WPS minima neighboring the region. If the identified region is 150-450 bp, we apply the same above median contiguous window approach, but only report those windows that are 50-150 bp. For score calculation of multiple windows derived from 150-450 bp regions, we set the neighboring minima to zero. We discard regions &lt;50 bp or &gt;450 bp.</p>
<p>Peak calling is implemented in <code>callPeaks.py</code> and expects WIG on STDIN:</p>
<div class="highlight highlight-source-shell"><pre><span class="pl-c"><span class="pl-c">#</span> Note that EXTREGION should be larger than REGION (by the maximum read length, i.e. 180) to prevent skipping alignments</span>
./samtools view -u -m 120 -M 180 BAMFILE.bam <span class="pl-smi">$EXTREGION</span> <span class="pl-k">|</span> ./FilterUniqueBAM.py -p <span class="pl-k">|</span> ./extractReadStartsFromBAM2Wig.py -p -r <span class="pl-smi">$REGION</span> -w 120 -c OFF -s OFF <span class="pl-k">|</span> ./callPeaks.py -s <span class="pl-k">&gt;</span> calls.bed

<span class="pl-c"><span class="pl-c">#</span> or from bigWig (http://hgdownload.cse.ucsc.edu/admin/exe/) and save as block-gzip compressed (http://www.htslib.org/doc/tabix.html) BED file:</span>

bigWigToWig -chrom=chr1 -start=12000000 -end=13000000 <span class="pl-smi">${SAMPLE}</span>.bw /dev/stdout <span class="pl-k">|</span> ./callPeaks.py -s <span class="pl-k">|</span> bgzip -c <span class="pl-k">&gt;</span> calls.bed.gz</pre></div>
<p>Again note the variables <code>EXTREGION</code> and <code>REGION</code> above. When piping reads from a region, we might miss the forward read of a read pair and the provided scripts usually only extract information from the first read of a pair. Thus, <code>EXTREGION</code> should include additional bases around the region (i.e. 200bp or another value that guarantees that all read insert sizes are included; here 180bp would be sufficient; e.g. if REGION is 1:1000-2000, EXTREGION should be 1:800-2200).</p>
<p>Bed files can be converted to bigBed using the above mentioned <a href="http://hgdownload.cse.ucsc.edu/admin/exe/">UCSC tools</a>. We uploaded bigBed (bb) files of our nucleosome calls to GEO for BH01, IH01, IH02, CH01, and CA01.</p>
<h2>
<a id="analysis-of-dhs-sites" class="anchor" href="#analysis-of-dhs-sites" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Analysis of DHS sites</h2>
<p>DHS peaks for 349 primary tissue and cell line samples were downloaded from <a href="http://www.uwencode.org/proj/Science_Maurano_Humbert_et_al/data/all_fdr0.05_hot.tgz">http://www.uwencode.org/proj/Science_Maurano_Humbert_et_al/data/all_fdr0.05_hot.tgz</a>. Samples derived from fetal tissues (233 of 349) were removed from the analysis as they behaved inconsistently within tissue type, possibly because of unequal representation of cell types within each tissue sample. 116 DHS callsets from a variety of cell lineages were retained for analysis. For the midpoint of each DHS peak in a particular set, the nearest upstream and downstream calls in the CH01 callset were identified, and the distance between the centers of those two calls was calculated. The distribution of all such distances was visualized for each DHS peak callset using a smoothed density estimate calculated for distances between 0 and 500 bp.</p>
<p>See the folder <code>DHS</code> for more details and supporting files.</p>
<h2>
<a id="analysis-of-ab-compartments" class="anchor" href="#analysis-of-ab-compartments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Analysis of A/B compartments</h2>
<p>We downloaded A/B compartment calls with 100 kb resolution from <a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63525">GSE63525</a>, specifically the file <code>GSE63525_GM12878_subcompartments.bed.gz</code>. A/B segmentation calls were compared to peak spacing within equivalent 100 kb genomic windows.  In a separate analysis, peak density was compared to windowed GC content in 10 kb bins.</p>
<p>See the folder <code>peak_density</code> for more details and supporting files.</p>
<h2>
<a id="annotation-of-tfbss-and-genomic-features" class="anchor" href="#annotation-of-tfbss-and-genomic-features" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Annotation of TFBSs and genomic features</h2>
<p>We downloaded clustered FIMO (motif-based) intervals (<a href="http://www.ncbi.nlm.nih.gov/pubmed/21330290">Grant et al., 2011</a>; <a href="http://www.ncbi.nlm.nih.gov/pubmed/22955828">Maurano et al., 2012</a>, <a href="http://www.uwencode.org/proj/Maurano_et_al_func_var/">http://www.uwencode.org/proj/Maurano_et_al_func_var/</a>) defining a set of computationally predicted TFBSs (<code>hg19.taipale.1e-4.starch</code>). For a subset of clustered TFs (AP-2-2, AP-2, CTCF_Core-2, E2F-2, EBF1, Ebox-CACCTG, Ebox, ESR1, ETS, MAFK, MEF2A-2, MEF2A, MYC-MAX, PAX5-2, RUNX2, RUNX-AML, STAF-2, TCF-LEF, YY1), we retained only predicted TFBSs that overlap with ENCODE ChIP-seq peaks (TfbsClusteredV3 set downloaded from <a href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeRegTfbsClustered/">http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeRegTfbsClustered/</a>). Details on download and basic processing of these data sets are available in <code>data/Maurano_et_al_func_var/README</code> and <code>data/ENCODE_TfbsClusteredV3/README</code>.</p>
<p>Genomic coordinates of transcription start sites, transcription end sites, start codons, and splice donor and acceptor sites were obtained from Ensembl Build version 75 (ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/Homo_sapiens.GRCh37.75.gtf.gz, details are available in <code>data/Ensembl_v75/README</code>).</p>
<h2>
<a id="filtering-active-ctcf-sites" class="anchor" href="#filtering-active-ctcf-sites" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Filtering active CTCF sites</h2>
<p>CTCF sites first included clustered FIMO binding site predictions (described above). This set was intersected with ENCODE ChIP-seq peaks (TfbsClusteredV3, described above), and then further intersected with a set of CTCF binding sites experimentally observed to be active across 19 tissues (Wang et al., 2012; details on the download of this data set are available in <code>data/Wang_et_al_CTCF/README</code>), to produce three increasingly stringent sets. For each CTCF site, distances between each of 20 flanking nucleosomes  (10 upstream and 10 downstream) were calculated.</p>
<p>The mean S-WPS and L-WPS at each position relative to the center of the CTCF binding motifs were calculated within bins defined by spacing between -1 and +1 nucleosomes (&gt;160 bp, 161-200 bp, 201-230 bp, 231-270 bp, 271-420 bp, 421-460 bp, and &gt;460 bp). In figures, CTCF binding sites are shifted such that the zero coordinate on the x-axis is the center of its 52 bp binding footprint (<a href="http://www.ncbi.nlm.nih.gov/pubmed/24614316">Ong and Corces, 2014</a>).</p>
<h2>
<a id="overlaying-wps-at-aligned-genomic-features" class="anchor" href="#overlaying-wps-at-aligned-genomic-features" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Overlaying WPS at aligned genomic features</h2>
<p>WPS values for a specific data set (in most cases CH01) and the corresponding simulation were extracted for each position in a 5 kb window around the start coordinate of each TFBS, and was aggregated within each TF cluster. The mean WPS of the first and last 500 bp (which is predominantly flat and represents a mean offset) of the 5 kb window was subtracted from the original WPS at each position. For L-WPS only, a sliding window mean is calculated using a 200 bp window and subtracted from the original signal. Finally, the corrected WPS profile for the simulation is subtracted from the corrected WPS profile of the data set to correct for signal that is a product of fragment length and ligation bias. This final profile is plotted and termed the Adjusted WPS.</p>
<p>We provide Python and R scripts for extracting and plotting WPS overlays from block-gzip compressed WIG files. Such files are generated by <code>extractReadStartsFromBAM2Wig.py</code> or <code>normalizeWPSwigs.py</code> and need to be tabix-indexed to retrieve specific genomic regions using the tabix routines of <a href="http://pysam.readthedocs.org/en/latest/api.html">pysam</a>.</p>
<p>###Motif-based TF binding site predictions</p>
<p>We outline how S-WPS and L-WPS plots were generated from clustered FIMO (motif-based) sites in a README in <code>WPS_overlays/Maurano_et_al_TFclusters</code>.</p>
<p>###Motif-based TF binding sites overlapped with ENCODE ChIP-seq peaks</p>
<p>Please find a README in <code>WPS_overlays/Maurano_et_al_plus_ChIP</code> which outlines how S-WPS and L-WPS plots were generated for a subset of clustered FIMO (motif-based) sites overlapped with ENCODE ChIP-seq peaks of AP-2-2, AP-2, CTCF_Core-2, E2F-2, EBF1, Ebox-CACCTG, Ebox, ESR1, ETS, MAFK, MEF2A-2, MEF2A, MYC-MAX, PAX5-2, RUNX2, RUNX-AML, STAF-2, TCF-LEF, and YY1.</p>
<p>###Active CTCF from ChIP-seq in 19 cell-lines</p>
<p><code>WPS_overlays/CTCF_19CellLines</code> has a README file which outlines how S-WPS and L-WPS plots were generated for this subset of CTCF sites.</p>
<p>###Adjusted WPS around genic features</p>
<p>A README describing how the adjusted WPS was extracted and overlayed around TSS, TSE, splice acceptor/donor, and start/stop codon can be found in <code>WPS_overlays/genicFeatures</code>.</p>
<p>###Adjusted WPS around genic features in five NB-4 expression bins</p>
<p>The file <code>WPS_overlays/genicFeatures_by_expression/README</code> describes how adjusted WPS was extracted and overlayed around TSS, TSE, splice acceptor/donor, and start/stop codon for the five NB-4 expression bins and how the genes in each bin were defined.</p>
<h2>
<a id="gene-expression-analysis" class="anchor" href="#gene-expression-analysis" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Gene expression analysis</h2>
<p>###Human Protein Atlas</p>
<p>FPKM gene expression (GE) values measured for 20,344 Ensembl gene identifiers in 44 human cell lines and 32 primary tissues by the Human Protein Atlas (<a href="http://www.ncbi.nlm.nih.gov/pubmed/25613900">Uhlén et al., 2015</a>) was downloaded from <a href="http://www.proteinatlas.org/download/rna.csv.zip">http://www.proteinatlas.org/download/rna.csv.zip</a>. Genes with 3 or more non-zero expression values were retained (n=19,378 genes). The GE data set is provided with one decimal precision for the FPKM values. Thus, a zero GE value (0.0) indicates expression in the interval [0, 0.05) Unless otherwise noted, we set the minimum GE value to 0.04 FPKM before log2-transformation..</p>
<p>###Fast Fourier transformation (FFT) and smoothing of trajectories</p>
<p>We use parameters to smooth (3 bp Daniell smoother; moving average giving half weight to the end values) and de-trend the data (i.e. subtract the mean of the series and remove a linear trend). A recursive time series filter implemented in R was used to remove high frequency variation from trajectories. 24 filter frequencies (1/seq(5,100,4)) were used, and the first 24 values of the trajectory were taken as init values. The 24-value shift in the resulting trajectories was corrected by repeating the last 24 values of the trajectory.</p>
<p>###FFT intensity correlation with expression</p>
<p>L-WPS was used to calculate periodograms of genomic regions using Fast Fourier Transform (FFT, spec.pgram in R) with frequencies between 1/500 and 1/100 bases. Intensity values for the 120-280 bp frequency range were determined from smooth FFT periodograms. S-shaped Pearson correlation between GE values and FFT intensities was observed around the major inter-nucleosome distance peak, along with a pronounced negative correlation in the 193-199 bp frequency range. The mean intensity in this frequency range was correlated with the average intensity with log2-transformed GE values for downstream analysis.</p>
<p>###Running the analysis</p>
<div class="highlight highlight-source-shell"><pre>
zcat transcriptAnno-GRCh37.75.tsv.gz <span class="pl-k">|</span> awk <span class="pl-s"><span class="pl-pds">'</span>BEGIN{ FS="\t"; OFS="\t" }{ if ($5 == "+") { print $1,$2,$3-10000,$3,$5 } else { print $1,$2,$4,$4+10000,$5 } }<span class="pl-pds">'</span></span> <span class="pl-k">&gt;</span> transcriptAnno-GRCh37.75.upstream.tsv                                                   
zcat transcriptAnno-GRCh37.75.tsv.gz <span class="pl-k">|</span> awk <span class="pl-s"><span class="pl-pds">'</span>BEGIN{ FS="\t"; OFS="\t" }{ if ($5 == "+") { print $1,$2,$4,$4+10000,$5 } else { print $1,$2,$3-10000,$3,$5 } }<span class="pl-pds">'</span></span> <span class="pl-k">&gt;</span> transcriptAnno-GRCh37.75.downstream.tsv                                                 
zcat transcriptAnno-GRCh37.75.tsv.gz <span class="pl-k">|</span> awk <span class="pl-s"><span class="pl-pds">'</span>BEGIN{ FS="\t"; OFS="\t" }{ if ($5 == "+") { print $1,$2,$3-1,$3-1+10000,$5 } else { print $1,$2,$4-1-10000,$4-1,$5 } }<span class="pl-pds">'</span></span> <span class="pl-k">&gt;</span> transcriptAnno-GRCh37.75.body.tsv

<span class="pl-c"><span class="pl-c">#</span> Extract counts for a sample</span>
mkdir -p body/<span class="pl-smi">$SAMPLE</span>
./extractReadStartsFromBAM_Region_WPS.py --minInsert=120 --maxInsert=180 -i transcriptAnno-GRCh37.75.body.tsv -o <span class="pl-s"><span class="pl-pds">'</span>body/$SAMPLE/block_%s.tsv.gz<span class="pl-pds">'</span></span> BAMFILE.bam

<span class="pl-c"><span class="pl-c">#</span> Run FFT &amp; convert to summary tbl</span>
mkdir -p /tmp/body/<span class="pl-smi">$SAMPLE</span>/fft
( <span class="pl-c1">cd</span> body/<span class="pl-smi">$SAMPLE</span>/counts<span class="pl-k">;</span> ls block_<span class="pl-k">*</span>.tsv.gz ) <span class="pl-k">|</span> xargs -n 500 Rscript fft_path.R body/<span class="pl-smi">$SAMPLE</span>/counts /tmp/body/<span class="pl-smi">$SAMPLE</span>/fft
mkdir -p body/fft_summaries/
./convert_files.py -a transcriptAnno-GRCh37.75.body.tsv -t /tmp/ -r  -p body -i <span class="pl-smi">$SAMPLE</span>
rm -fR /tmp/body/<span class="pl-smi">$SAMPLE</span>/fft

<span class="pl-c"><span class="pl-c">#</span> Correlate the intensities with the expression data and generate PDFs in R</span>
R --vanilla --quiet <span class="pl-k">&lt;</span> plots.R</pre></div>
        </section>

        <footer>
          Analysis of epigenetic signals captured by fragmentation patterns of cell-free DNA is maintained by <a href="https://github.com/shendurelab">shendurelab</a><br>
          This page was generated by <a href="https://pages.github.com">GitHub Pages</a>. Tactile theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.
        </footer>

        
      </div>
    </div>
  </body>
</html>
